File: react-school-frontend/src/services/CourseService.ts
import api from "@/services/api";

type RawOption = {
  id?: number | string | null;
  name?: string | null;
  title?: string | null;
  label?: string | null;
  value?: string | null;
};

type RawTeacher = RawOption & {
  fullName?: string | null;
  userT?: {
    id?: number | string | null;
    name?: string | null;
    fullName?: string | null;
    email?: string | null;
  } | null;
};

type RawCourse = {
  id?: number | string | null;
  title?: string | null;
  description?: string | null;
  published?: boolean | null;
  createdAt?: string | null;
  subjectName?: string | null;
  yearName?: string | null;
  streamName?: string | null;
  substreamName?: string | null;
  year?: RawOption | string | null;
  stream?: RawOption | string | null;
  substream?: RawOption | string | null;
  subject?: RawOption | string | null;
  teacher?: RawTeacher | null;
  lesson?: unknown[] | null;
  lessons?: unknown[] | null;
  lessonsCount?: number | null;
  quizzesCount?: number | null;
  enrollments?: unknown[] | null;
  enrolledCount?: number | null;
};

type RawLesson = {
  id?: number | string | null;
  title?: string | null;
  content?: string | null;
  orderNumber?: number | null;
  courseId?: number | string | null;
  videoUrl?: string | null;
  pdfUrl?: string | null;
  resource?: {
    id?: number | string | null;
    videoUrl?: string | null;
    pdfUrl?: string | null;
  } | null;
  courses?: {
    id?: number | string | null;
  } | null;
};

export interface Course {
  id: number;
  title: string;
  description: string;
  published: boolean;
  createdAt: string;
  subject: string;
  year: string;
  stream: string;
  substream: string;
  teacherId: number | null;
  teacherName: string;
  lessonsCount: number;
  quizzesCount: number;
  enrolledCount: number;
}

export interface Lesson {
  id: number;
  title: string;
  content: string;
  orderNumber: number;
  courseId: number | null;
  videoUrl: string;
  pdfUrl: string;
}

export interface CoursePayload {
  title: string;
  description?: string;
  published?: boolean;
  year?: string;
  streamId?: number;
  substreamId?: number;
  subjectId?: number;
  teacherId?: number;
}

export interface LessonPayload {
  title: string;
  content?: string;
  orderNumber?: number;
  courseId?: number;
}

const getLabel = (value: RawOption | string | null | undefined) => {
  if (!value) return "";
  if (typeof value === "string") return value;

  return (
    value.name ??
    value.title ??
    value.label ??
    value.value ??
    (value.id !== undefined && value.id !== null ? String(value.id) : "")
  );
};

const toNumber = (value: number | string | null | undefined) => {
  if (typeof value === "number") return value;
  if (typeof value === "string" && value.trim() !== "") {
    const parsed = Number(value);
    return Number.isNaN(parsed) ? 0 : parsed;
  }
  return 0;
};

const normalizeCourse = (course: RawCourse): Course => {
  const teacherId =
    course.teacher?.id !== undefined && course.teacher?.id !== null
      ? toNumber(course.teacher.id)
      : course.teacher?.userT?.id !== undefined && course.teacher?.userT?.id !== null
        ? toNumber(course.teacher.userT.id)
      : null;

  return {
    id: toNumber(course.id),
    title: course.title ?? "Untitled course",
    description: course.description ?? "",
    published: Boolean(course.published),
    createdAt: course.createdAt ?? "",
    subject: course.subjectName ?? getLabel(course.subject),
    year: course.yearName ?? getLabel(course.year),
    stream: course.streamName ?? getLabel(course.stream),
    substream: course.substreamName ?? getLabel(course.substream ?? course.subject),
    teacherId,
    teacherName:
      course.teacher?.fullName ??
      course.teacher?.userT?.fullName ??
      course.teacher?.name ??
      course.teacher?.userT?.name ??
      course.teacher?.title ??
      "Teacher",
    lessonsCount: course.lessonsCount ??
      (Array.isArray(course.lesson)
        ? course.lesson.length
        : Array.isArray(course.lessons)
          ? course.lessons.length
          : 0),
    quizzesCount: course.quizzesCount ?? 0,
    enrolledCount: course.enrolledCount ?? (Array.isArray(course.enrollments) ? course.enrollments.length : 0),
  };
};

const normalizeLesson = (lesson: RawLesson): Lesson => ({
  id: toNumber(lesson.id),
  title: lesson.title ?? "Untitled lesson",
  content: lesson.content ?? "",
  orderNumber: lesson.orderNumber ?? 0,
  videoUrl: lesson.resource?.videoUrl ?? lesson.videoUrl ?? "",
  pdfUrl: lesson.resource?.pdfUrl ?? lesson.pdfUrl ?? "",
  courseId:
    lesson.courseId !== undefined && lesson.courseId !== null
      ? toNumber(lesson.courseId)
      : lesson.courses?.id !== undefined && lesson.courses?.id !== null
      ? toNumber(lesson.courses.id)
      : null,
});

const normalizeCourseList = (data: unknown) => {
  const list =
    Array.isArray((data as { content?: RawCourse[] })?.content)
      ? (data as { content: RawCourse[] }).content
      : Array.isArray(data)
        ? (data as RawCourse[])
        : [];

  return list.map(normalizeCourse);
};

export const getCourses = async (page = 0, size = 10) => {
  const { data } = await api.get("/courses", { params: { page, size } });
  return normalizeCourseList(data);
};

export const getCourseById = async (id: number) => {
  const { data } = await api.get(`/courses/${id}`);
  return normalizeCourse(data);
};

export const createCourse = async (course: CoursePayload) => {
  const { data } = await api.post("/courses", course);
  return normalizeCourse(data);
};

export const updateCourse = async (id: number, course: CoursePayload) => {
  await api.put(`/courses/${id}`, course);
};

export const deleteCourse = async (id: number) => {
  await api.delete(`/courses/${id}`);
};

export const publishCourse = async (id: number) => {
  const { data } = await api.post(`/courses/${id}/publish`);
  return normalizeCourse(data);
};

export const unpublishCourse = async (id: number) => {
  const { data } = await api.post(`/courses/${id}/unpublish`);
  return normalizeCourse(data);
};

export const getPopularCourses = async (
  studentId: number,
  page = 0,
  size = 10
) => {
  const { data } = await api.get(`/students/${studentId}/courses/popular`, {
    params: { page, size },
  });

  return normalizeCourseList(data);
};

export const getRecommendedCourses = async (studentId: number, page = 0, size = 10) => {
  const { data } = await api.get(`/students/${studentId}/courses/recommended`, {
    params: { page, size },
  });

  return normalizeCourseList(data);
};

export const getLatestCourses = async (studentId: number, page = 0, size = 10) => {
  const { data } = await api.get(`/students/${studentId}/courses/latest`, {
    params: { page, size },
  });

  return normalizeCourseList(data);
};

/** Paginated published courses with optional filters (years.id / stream / subject). */
export const discoverCourses = async (
  page = 0,
  size = 12,
  filters?: { streamId?: number | null; subjectId?: number | null; yearId?: number | null }
) => {
  const params: Record<string, number> = { page, size };
  if (filters?.streamId != null) params.streamId = filters.streamId;
  if (filters?.subjectId != null) params.subjectId = filters.subjectId;
  if (filters?.yearId != null) params.yearId = filters.yearId;
  const { data } = await api.get("/courses/discover", { params });
  return normalizeCourseList(data);
};

export const getLessonsByCourse = async (courseId: number) => {
  const { data } = await api.get(`/courses/${courseId}/lessons`);
  const list = Array.isArray(data) ? data : [];
  return list.map(normalizeLesson);
};

export const getLessonById = async (id: number) => {
  const { data } = await api.get(`/lessons/${id}`);
  return normalizeLesson(data);
};

export const createLesson = async (lesson: LessonPayload) => {
  if (lesson.courseId) {
    const { data } = await api.post(`/courses/${lesson.courseId}/lessons`, lesson);
    return normalizeLesson(data);
  }

  const { data } = await api.post("/lessons", lesson);
  return normalizeLesson(data);
};

export const updateLesson = async (lessonId: number, lesson: LessonPayload) => {
  const { data } = lesson.courseId
    ? await api.put(`/courses/${lesson.courseId}/lessons/${lessonId}`, lesson)
    : await api.put(`/lessons/${lessonId}`, { ...lesson, courseId: 0 });
  return normalizeLesson(data);
};

export const deleteLesson = async (lessonId: number, courseId?: number) => {
  if (courseId) {
    await api.delete(`/courses/${courseId}/lessons/${lessonId}`);
    return;
  }

  await api.delete(`/lessons/${lessonId}`);
};

File: spring-boot-school-backend/demo/src/main/java/com/educationonline/backend/services/CourseService.java
package com.educationonline.backend.services;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.stream.Collectors;

import jakarta.persistence.criteria.Predicate;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;

import com.educationonline.backend.dtos.CourseDto;
import com.educationonline.backend.dtos.StreamDto;
import com.educationonline.backend.dtos.SubstreamDto;
import com.educationonline.backend.dtos.TeacherCourseDTO;
import com.educationonline.backend.dtos.TeacherCourseOptionDto;
import com.educationonline.backend.dtos.TeacherPublicCourses;
import com.educationonline.backend.dtos.profiles.StudentProfileDto;
import com.educationonline.backend.entities.BadgeEventType;
import com.educationonline.backend.entities.Courses;
import com.educationonline.backend.entities.Enrollment;
import com.educationonline.backend.entities.Streams;
import com.educationonline.backend.entities.StudentLevel;
import com.educationonline.backend.entities.students;
import com.educationonline.backend.entities.subjects;
import com.educationonline.backend.entities.substream;
import com.educationonline.backend.entities.teachers;
import com.educationonline.backend.mappers.CourseMapper;
import com.educationonline.backend.mappers.LessonMapper;
import com.educationonline.backend.mappers.StreamMapper;
import com.educationonline.backend.mappers.SubstreamMapper;
import com.educationonline.backend.mappers.YearMapper;
import com.educationonline.backend.repositories.*;
import com.educationonline.backend.entities.Years;
import jakarta.transaction.Transactional;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.Setter;

@Getter
@Setter
@AllArgsConstructor


@Service
public class CourseService {

    private final StreamRepository streamRepository;
    private final CourseRepository courseRepository;
    private final LessonRepository lessonRepository;
    private final EnrollmentRepository enrollmentRepository;
    private final studentRepository studentRepository;
    private final teachersRepository teachersRepository;
    private final substreamRepository substreamRepository;
    private final yearRepo yearRepo;
    private final BadgeEngineService BadgeEngineService;
    private final PointsService pointsService;
    private final NotificationService notificationService;

    
    /*        TEACHER DASHBOARD 
         - createCourse()
         - updateCourse()
         - publishCourse()          validates course has lessons before publishing
         - unpublishCourse()
         - deleteCourse()
         - getCourseStudents()      list of enrolled students
         - getCourseStats()         total enrolled, avg progress, completed count
 
     */
    @Transactional
    public CourseDto createCourse(CourseDto course) {
        

        Courses newcourse = CourseMapper.toEntity(course);

// set year enum from yearId
if (course.getYearId() != null) {
    Years year = yearRepo.findById(course.getYearId())
        .orElseThrow(() -> new IllegalArgumentException("Year not found: " + course.getYearId()));
    newcourse.setYear(year.getYear()); 
}


         if (newcourse.getSubject() == null && newcourse.getTeacher() != null) {
             teachers teacher = teachersRepository.findById(newcourse.getTeacher().getId())
            .orElseThrow(() -> new IllegalStateException("Teacher not found"));

            if (teacher.getSubject().getId() != null) {
            subjects subject = new subjects();
            subject.setId(teacher.getSubject().getId());
            newcourse.setSubject(subject);
            }

        System.out.println("Set subject from teacher: " + teacher.getSubject().getId());

      }
        newcourse.setPublished(false); // course created but still not visible to students
        Courses saved = courseRepository.save(newcourse);
        return toCourseDto(saved);

    }

    @Transactional
    public CourseDto publishCourse(Long courseId) {
        Courses course = courseRepository.findById(courseId)
                .orElseThrow(() -> new IllegalStateException("Course not found"));
        if (lessonRepository.countByCoursesId(courseId) == 0) {
            throw new IllegalStateException("Cannot publish course without lessons");
        }
        course.setPublished(true);

        Courses saved = courseRepository.save(course);
        return toCourseDto(saved);
    }

    @Transactional // hide course but not delete it
    public CourseDto unpublishCourse(Long courseId) {
        Courses course =courseRepository.findById(courseId)
                .orElseThrow(() -> new IllegalStateException("Course not found"));
        long activeEnrollments = enrollmentRepository.countByCourses_IdAndStatus(
                courseId, Enrollment.EnrollmentStatus.ACTIVE);
        if (activeEnrollments > 0) {
            throw new IllegalStateException("Cannot unpublish a course while students are actively enrolled.");
        }
        course.setPublished(false);

        Courses saved = courseRepository.save(course);
        return toCourseDto(saved);
    }

    @Transactional
    public CourseDto updateCourse(Long courseId, CourseDto updatedCourse) {
        Courses course = courseRepository.findById(courseId)
                .orElseThrow(() -> new IllegalStateException("Course not found"));
        if (updatedCourse.getTitle() != null) {
            course.setTitle(updatedCourse.getTitle());
        }
          if (updatedCourse.getDescription()!=null) {
            course.setDescription(updatedCourse.getDescription());
          }   
          
          if (updatedCourse.getStudentLevel() != null) {
            course.setStudentLevel(
                StudentLevel.valueOf(updatedCourse.getStudentLevel())
        );
          }
        if (updatedCourse.getYearId() != null) {
    Years year = yearRepo.findById(updatedCourse.getYearId())
        .orElseThrow(() -> new RuntimeException("Year not found"));

course.setYear(year.getYear());
        
    }

    if (course.getStudentLevel() == StudentLevel.MIDDLE_SCHOOL) {
        course.setStream(null);
        course.setSubstream(null);
    } else {
        if (updatedCourse.getStreamId() != null) {
            Streams stream = streamRepository.findById(updatedCourse.getStreamId())
                    .orElseThrow(() -> new RuntimeException("Stream not found"));
            course.setStream(stream);
        }

        if (updatedCourse.getSubstreamId() != null) {
            substream sub = substreamRepository.findById(updatedCourse.getSubstreamId())
                    .orElseThrow(() -> new RuntimeException("Substream not found"));
            course.setSubstream(sub);
        }
    }
        Courses saved = courseRepository.save(course);
        return toCourseDto(saved);
    }

     @Transactional
    public void deleteCourse(Long courseId) {
        if (!courseRepository.existsById(courseId)) {
            throw new IllegalStateException("Course not found " );
        }
        courseRepository.deleteById(courseId);
    }

    public List<Enrollment> getAllStudentsInCourse(Long courseId) {
        return enrollmentRepository.findByCoursesId(courseId);
    }
/* 
public CourseStatsDTO getCourseStats(Long courseId) {
        List<Enrollment> enrollments = enrollmentRepository.findByCoursesId(courseId);
        long total     = enrollments.size();
        long completed = enrollments.stream()
                .filter(e -> e.getStatus() == Enrollment.EnrollmentStatus.COMPLETED).count();
        double avg     = enrollments.stream()
                .mapToInt(Enrollment::getProgressPercent).average().orElse(0);
        return new CourseStatsDTO(total, completed, avg);
    }
 */

    public CourseDto getCourseById(Long courseId) {
        Courses course = courseRepository.findById(courseId)
                .orElseThrow(() -> new IllegalStateException("Course not found"));
        return toCourseDto(course);
    }

    /** Published course only; used by public search / course view. */
    public CourseDto getPublishedCourseById(Long courseId) {

        Courses course = courseRepository.findByPublishedTrueAndId(courseId);
        return toCourseDto(course);
    }

    public List<CourseDto> getAllCoursesByTeacherId(Long teacherId) {

        List<Courses> courses = courseRepository.findByTeacherId(teacherId);
        return courses.stream().map(this::toCourseDto).toList();
    }

    public Long countCoursesByTeacherId(Long teacherId) {
        return courseRepository.countByTeacherId(teacherId);
    }

    /* ORGANIZED LEARNING + OPEN LEARNING - Student Dashboard
      - enrollStudent()          student subscribes to a course
      - getEnrolledCourses()     All courses a student is enrolled in
      - getCourseProgress()      enrollment + lesson progress details
      - completeCourse()         
      - dropCourse()   
     */
     
      @Transactional
      public Enrollment enrollStudent(Long studentId, Long courseId) {
        if (!studentRepository.existsById(studentId)) {
            throw new IllegalStateException("Student not found");
        }

        Courses course = courseRepository.findByPublishedTrueAndId(courseId);
        if (course == null) {
            throw new IllegalStateException("Course not found or not published");
        }

        var existing = enrollmentRepository.findByStudentIdAndCoursesId(studentId, courseId);
        if (existing.isPresent()) {
            Enrollment enrollment = existing.get();
            if (enrollment.getStatus() == Enrollment.EnrollmentStatus.DROPPED) {
                enrollment.setStatus(Enrollment.EnrollmentStatus.ACTIVE);
                enrollment.setDroppedAt(null);
                enrollment.setDropCount((enrollment.getDropCount() == null ? 0 : enrollment.getDropCount()) + 1);
                enrollment.setLastAccessedAt(LocalDateTime.now());
                Enrollment saved = enrollmentRepository.save(enrollment);
                notificationService.notifyEnrollment(course, studentRepository.findById(studentId)
                        .orElseThrow(() -> new IllegalStateException("Student not found")));
                return saved;
            }
            return enrollment;
        }
        
        students student = studentRepository.findById(studentId)
                .orElseThrow(() -> new IllegalStateException("Student not found"));

        Enrollment saved = enrollmentRepository.save(Enrollment.builder()
                .student(student)
                .courses(course)
                .progressPercent(0)
                .pointsEarned(0)
                .status(Enrollment.EnrollmentStatus.ACTIVE)
                .lastAccessedAt(LocalDateTime.now())
                .build());
        notificationService.notifyEnrollment(course, student);
        return saved;
    }

    //All enroled courses for a student
    public List<Enrollment> getEnrolledCourses(Long studentId) { 
        return enrollmentRepository.findByStudentId(studentId);
    }

    //get progress details for a specific course only if enrolled
    public Enrollment getCourseProgress(Long studentId, Long courseId) {
        return enrollmentRepository.findByStudentIdAndCoursesId(studentId, courseId)
                .orElseThrow(() -> new IllegalStateException(
                        "No enrollment found for student " + studentId + " in course " + courseId));
    }

     @Transactional
    public Enrollment completeCourse(Long studentId, Long courseId) {
        Enrollment enrollment = getCourseProgress(studentId, courseId);
        if (enrollment.getStatus() == Enrollment.EnrollmentStatus.COMPLETED) return enrollment; // idempotent
        enrollment.setStatus(Enrollment.EnrollmentStatus.COMPLETED);
        enrollment.setProgressPercent(100);
        enrollment.setCompletedAt(LocalDateTime.now());
        pointsService.addPoints(studentId, 50, "COURSE_COMPLETED");
        BadgeEngineService.handleEvent(
    studentId,
    BadgeEventType.COURSE_COMPLETED,
    Map.of("courseId", courseId)
);
        return enrollmentRepository.save(enrollment);
    }

    @Transactional
    public void dropCourse(Long studentId, Long courseId) {
        Enrollment enrollment = getCourseProgress(studentId, courseId);
        enrollment.setStatus(Enrollment.EnrollmentStatus.DROPPED);
        enrollmentRepository.save(enrollment);
    }
/* 
    @Transactional
public Enrollment reEnroll(Long studentId, Long courseId, boolean continueProgress) {

    Enrollment existing = getCourseProgress(studentId, courseId);

    if (existing.getStatus() != Enrollment.EnrollmentStatus.DROPPED) {
        throw new IllegalStateException("Student is not dropped from this course.");
    }

    // Option 1 — continue from where they stopped
    existing.setStatus(Enrollment.EnrollmentStatus.ACTIVE);
    existing.setDroppedAt(null);
    existing.setDropCount(existing.getDropCount() + 1);

    if (continueProgress) {
    
    // record when they came back
    existing.setLastAccessedAt(LocalDateTime.now());

    // 2. recalculate progress percent based on actual completed lessons(in case lessons were added to the course while they were dropped)
    long totalLessons     = lessonRepository.countByCoursesId(courseId);
    long completedLessons = existing.getLessonProgresses()
            .stream()
            .filter(LessonProgress::getCompleted)
            .count();

    int recalculatedPercent = totalLessons == 0 ? 0
            : (int) ((completedLessons * 100) / totalLessons);

    existing.setProgressPercent(recalculatedPercent);

    // 3. keep all lesson progress exactly as it was —
    //    completed lessons stay completed,
    //    video positions are preserved so student resumes where they paused
    //    nothing else to change h
    }else{
    

    
    
     existing.setProgressPercent(0);
     existing.setPointsEarned(0);
      existing.setCompletedAt(null);}
    return enrollmentRepository.save(existing);
}*/
public List<Courses> getCourseByStream(Long streamId){

    return courseRepository.findByPublishedTrueAndStreamId(streamId);
}

public List<Courses> getCourseBySubject(Long subjectId){
    return courseRepository.findByPublishedTrueAndSubjectId(subjectId);
}

public List<Courses> getCourseByYear(String year){
    Streams stream = streamRepository.findByYear(year)
            .stream()
            .findFirst()
            .orElseThrow(() -> new IllegalStateException("Stream not found for year: " + year));
    return courseRepository.findByPublishedTrueAndStreamId(stream.getId());
}

public Page<CourseDto> getAllPublishedCourses(int page, int size) {
    Pageable pageable = PageRequest.of(page, size, Sort.by("createdAt").descending());
    
    Page<Courses> courses = courseRepository.findByPublishedTrue(pageable);
    return courses.map(this::toCourseDto);
}
// related with the endpoint /teachers/{teacherId}/courses?page=&size= to get only published courses with pagination for the student dashboard
@Transactional
public TeacherPublicCourses getPublishedCoursesByTeacher(Long teacherId, int page, int size) {
    teachersRepository.findById(teacherId)
        .orElseThrow(() -> new IllegalStateException("Teacher not found"));

    Pageable pageable = PageRequest.of(page, size, Sort.by("createdAt").descending());
    Page<Courses> coursePage = courseRepository.findByTeacherIdAndPublishedTrue(teacherId, pageable);
    
     Page<TeacherCourseDTO> dtoPage = coursePage.map(c -> {
        // Convert StudentLevel enum to String
        String studentLevelStr = c.getStudentLevel() != null 
            ? c.getStudentLevel().name()  
            : null;
            
        return new TeacherCourseDTO(
            c.getId(),
            c.getTitle(),
            c.getDescription(),
            studentLevelStr,              
            c.getYear() != null ? c.getYear().getValue() : null,
            c.getStream() != null ? c.getStream().getStreamType().getNamestream() : null,
            c.getSubstream() != null ? c.getSubstream().getNameSubstream() : null,
            c.isPublished(),
            c.getTeacher() != null ? c.getTeacher().getUserT().getName() : "Teacher",
            c.getCreatedAt() != null ? c.getCreatedAt().toString() : "",
            c.getLesson() != null ? c.getLesson().size() : 0,
            c.getEnrollments() != null ? c.getEnrollments().size() : 0
        );
    });
    
    return new TeacherPublicCourses(dtoPage);
}

    //FOR CREATION COURSES OPTIONS
    /* 
    public List<YearDto> getByStudentLevel(StudentLevel level) {

    List<Years> years =yearRepo.findByYearType(level);
        return years
            .stream()
            .map(YearMapper::toDto)
            .toList();
}*/

public List<StreamDto> getStreamsByYear(Long yearId){
    List<Streams> streams = streamRepository.findByYearId(yearId);

    return streams.stream().map(StreamMapper::toDto).toList();
}

public List<SubstreamDto> getSubstream(Long yearId, Long streamId){
    List<substream> substreams = substreamRepository.findByStreamsId( streamId);

    return substreams.stream().map(SubstreamMapper::toDto).toList();
}


public List<TeacherCourseOptionDto> getTeacherCourseOptions(Long teacherId) {

    return courseRepository.findByTeacherId(teacherId)
            .stream()
            .map(c -> new TeacherCourseOptionDto(
                    c.getId(),
                    c.getTitle()
            ))
            .toList();
}




public StudentProfileDto getStudentProfile(Long studentId) {
    students student = studentRepository.findById(studentId)
            .orElseThrow(() -> new IllegalStateException("Student not found"));

    Streams stream = student.getStreams();

    return new StudentProfileDto(
            student.getStudentLevel(),
            stream != null ? stream.getId() : null,
            stream != null && stream.getYear() != null ? stream.getYear().getId() : null,
            student.getSubstream() != null ? student.getSubstream().getId() : null
    );
            
}
public Page<CourseDto> getPopularCourses(Long studentId, int page, int size) {
    Pageable pageable = PageRequest.of(page, size);

    studentRepository.findById(studentId)
            .orElseThrow(() -> new IllegalStateException("Student not found"));

    Page<Courses> courses = courseRepository.findPopularCourses(pageable);
    return courses.map(this::toCourseDto);


}

public Page<CourseDto> getRecomendedCourses(Long studentId, int page, int size) {
    
    studentRepository.findById(studentId)
            .orElseThrow(() -> new IllegalStateException("Student not found"));
    Pageable pageable = PageRequest.of(page, size);
    StudentProfileDto profile = getStudentProfile(studentId);
    if (profile.studentLevel() == null) {
        return Page.empty(pageable);
    }

    // High school: filter by stream + year + substream
    if (profile.streamId() != null && profile.yearId() != null) {
        return courseRepository.findRecomendedCourses(
            profile.studentLevel(),
            profile.streamId(),
            profile.yearId(),
            profile.substreamId(),
            studentId,
            pageable
        ).map(this::toCourseDto);
    }

    // Middle school: filter by year from enrollments + substream (no stream)
    Years.yearValue studentYear = findStudentYearFromEnrollments(studentId);
    if (studentYear == null) {
        return Page.empty(pageable);
    }
    return courseRepository.findRecomendedCoursesMiddleSchool(
        profile.studentLevel(),
        studentYear,
        profile.substreamId(),
        studentId,
        pageable
    ).map(this::toCourseDto);
}

private Years.yearValue findStudentYearFromEnrollments(Long studentId) {
    List<Enrollment> enrollments = enrollmentRepository.findByStudentId(studentId);
    return enrollments.stream()
        .map(e -> e.getCourses())
        .filter(Objects::nonNull)
        .map(Courses::getYear)
        .filter(Objects::nonNull)
        .findFirst()
        .orElse(null);
}

public Page<CourseDto> getLatestCourses( Long studentId, int page, int size) {
    Pageable pageable = PageRequest.of(page, size, Sort.by("createdAt").descending());

    studentRepository.findById(studentId)
            .orElseThrow(() -> new IllegalStateException("Student not found"));

    LocalDateTime since = LocalDateTime.now().minusDays(5);
    return courseRepository.findLatestCoursesSince(since, pageable)
            .map(this::toCourseDto);
}

/** Optional filters; all-null returns all published (paged like catalog). */
public Page<CourseDto> discoverPublishedCourses(Long streamId, Long subjectId, Long yearTableId,
        int page, int size) {
    Pageable pageable = PageRequest.of(page, size, Sort.by("createdAt").descending());

    Years.yearValue yearFromPicker = yearTableId == null ? null
            : yearRepo.findById(yearTableId).map(Years::getYear).orElse(null);

    Years.yearValue yearFromStream = streamId == null ? null
            : streamRepository.findById(streamId)
                    .map(Streams::getYear)
                    .map(Years::getYear)
                    .orElse(null);

    final Years.yearValue yearFilter = yearFromPicker != null ? yearFromPicker : yearFromStream;

    Specification<Courses> spec = (root, query, cb) -> {
        List<Predicate> parts = new ArrayList<>();
        parts.add(cb.isTrue(root.get("published")));

        if (subjectId != null) {
            parts.add(cb.equal(root.get("subject").get("id"), subjectId));
        }

        if (streamId != null) {
            if (yearFilter != null) {
                Predicate matchStream = cb.equal(root.get("stream").get("id"), streamId);
                Predicate middleNoStream = cb.and(
                        cb.isNull(root.get("stream")),
                        cb.equal(root.get("year"), yearFilter),
                        cb.equal(root.get("studentLevel"), StudentLevel.MIDDLE_SCHOOL));
                parts.add(cb.or(matchStream, middleNoStream));
            } else {
                parts.add(cb.equal(root.get("stream").get("id"), streamId));
            }
        } else if (yearFilter != null) {
            parts.add(cb.equal(root.get("year"), yearFilter));
        }

        query.distinct(true);
        return cb.and(parts.toArray(Predicate[]::new));
    };

    return courseRepository.findAll(spec, pageable).map(this::toCourseDto);
}

private CourseDto toCourseDto(Courses course) {
    CourseDto dto = CourseMapper.toDto(course);

    if (course == null || dto == null || course.getYear() == null) {
        return dto;
    }

    yearRepo.findByYear(course.getYear())
            .map(Years::getId)
            .ifPresent(dto::setYearId);

    return dto;
}


}
